# Исследование надежности заемщиков

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Описание-проекта" data-toc-modified-id="Описание-проекта-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Описание проекта</a></span></li><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обзор данных</a></span></li><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Подготовка данных</a></span><ul class="toc-item"><li><span><a href="#Заполнение-пропусков" data-toc-modified-id="Заполнение-пропусков-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Заполнение пропусков</a></span></li><li><span><a href="#Проверка-данных-на-аномалии-и-исправления" data-toc-modified-id="Проверка-данных-на-аномалии-и-исправления-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Проверка данных на аномалии и исправления</a></span></li><li><span><a href="#Изменение-типов-данных" data-toc-modified-id="Изменение-типов-данных-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Изменение типов данных</a></span></li><li><span><a href="#Удаление-дубликатов" data-toc-modified-id="Удаление-дубликатов-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Удаление дубликатов</a></span></li><li><span><a href="#Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма" data-toc-modified-id="Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма</a></span></li><li><span><a href="#Категоризация-дохода" data-toc-modified-id="Категоризация-дохода-3.6"><span class="toc-item-num">3.6&nbsp;&nbsp;</span>Категоризация дохода</a></span></li><li><span><a href="#Категоризация-целей-кредита" data-toc-modified-id="Категоризация-целей-кредита-3.7"><span class="toc-item-num">3.7&nbsp;&nbsp;</span>Категоризация целей кредита</a></span></li></ul></li><li><span><a href="#Построение-зависимостей" data-toc-modified-id="Построение-зависимостей-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Построение зависимостей</a></span><ul class="toc-item"><li><span><a href="#Количество-детей-и-возврат-кредита-в-срок" data-toc-modified-id="Количество-детей-и-возврат-кредита-в-срок-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Количество детей и возврат кредита в срок</a></span></li><li><span><a href="#Семейное-положение-и-возврат-кредита-в-срок" data-toc-modified-id="Семейное-положение-и-возврат-кредита-в-срок-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Семейное положение и возврат кредита в срок</a></span></li><li><span><a href="#Уровень-дохода-и-возврат-кредита-в-срок" data-toc-modified-id="Уровень-дохода-и-возврат-кредита-в-срок-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Уровень дохода и возврат кредита в срок</a></span></li><li><span><a href="#Цель-кредита-и-возврат-кредита-в-срок" data-toc-modified-id="Цель-кредита-и-возврат-кредита-в-срок-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Цель кредита и возврат кредита в срок</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Общий вывод</a></span></li></ul></div>

## Описание проекта

Кредитный отдел банка обратился с вопросом оценки способности потенциального заемщика вернуть кредит. Для этого он хочет, чтобы была проанализирована предоставленная статистика о платежеспособности существующих клиентов.

**Цель исследования** — выяснить, влияют ли на возврат кредита в срок такие факторы как:

* количество детей;

* семейное положение;

* уровень дохода;

* цель кредита.

**Ход исследования**

Форма предоставленных данных не была согласована заранее, поэтому понадобится обзор данных и их подготовка, в процессе которой проверим данные на ошибки и при необходимости исправим их.  
Разделим данные на категории, выявим закономерности в разных категориях и построим сводные таблицы.  
Таким образом, исследование будет проводиться в три этапа:  
1. Обзор данных.
2. Предобработка данных.
3. Построение зависимостей.


## Обзор данных

In [1]:
#Импортируем pandas и откроем предоставленную таблицу (варианты для локальной и удаленной работы)
import pandas as pd
try:
    stat = pd.read_csv('/datasets/data.csv')
except:
    stat = pd.read_csv('dataset_project2.csv')

In [2]:
#Просмотрим общую информацию о фрейме.
stat.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [3]:
#Просмотрим первые 10 строк фрейма.
stat[:10]

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

In [4]:
#Посмотрим названия столбцов, чтобы посмотреть, нет ли ошибок в наименовании столбцов.
stat.columns

Index(['children', 'days_employed', 'dob_years', 'education', 'education_id',
       'family_status', 'family_status_id', 'gender', 'income_type', 'debt',
       'total_income', 'purpose'],
      dtype='object')

Как видно из общей информации о предоставленном датафрейме, данные разделены на 12 столбцов, в наименовании которых нет ошибок. Согласно предоставленной пояснительной записке к данным, соответствие названия столбцов и их содержания следующее:
* `children` - количество детей в семье;
* `days_employed` - общий трудовой стаж в днях;
* `dob_years` - возраст клиента в годах;
* `education` - уровень образования клиента;
* `education_id` - идентификатор уровня образования;
* `family_status` - семейное положение;
* `family_status_id` - идентификатор семейного положения;
* `gender` - пол клиента;
* `income_type` - тип занятости;
* `debt` - имел ли задолженность по возврату кредитов;
* `total_income` - ежемесячный доход;
* `purpose` - цель получения кредита.

Из общей информации и части таблицы видны проблемы в данных:
* в столбцах `days_employed` и `total_income` имеются пропущенные значения;
* в столбце `days_employed` встречаются отрицательные значения;
* в столбце `education` значения записаны в разных регистрах.

**Выводы:**  
В каждой строке таблицы - данные о заемщике и взятом им кредите. Часть колонок описывают кредит: были ли просрочки и на какие цели он взят. Остальные колонки описывают заемщика: сколько лет, стаж, размер зарплаты, количество детей и т.д.  

Данных достаточно, чтобы можно было выявить требуемые закономерности. Однако, проблемы в данных могут помешать их обработать.  

Чтобы перейти к построению зависимостей, нужно устранить проблемы в данных.

## Подготовка данных

### Заполнение пропусков

На предыдущем шаге было определено, что в столбцах `days_employed` и `total_income` есть пропущенные значения.  
Определим, сколько их.

In [5]:
#Проверим, в каких стоблцах есть пропуски и сколько их
stat.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

Видно, что пропуски есть только в этих колонках. И в каждом из них по 2174 пропуска. Странно, что в разных столбцах одинаковое количество пропусков. Проверим, возможно, пропуски в этих столбцах встречаются в одних и тех же строках.

In [6]:
#выберем строки, в которых в обоих столбцах (days_employed и total_income) значения пропущены.
stat[stat['days_employed'].isna() & stat['total_income'].isna()]

children  days_employed  dob_years education  education_id  \
12            0            NaN         65   среднее             1   
26            0            NaN         41   среднее             1   
29            0            NaN         63   среднее             1   
41            0            NaN         50   среднее             1   
55            0            NaN         54   среднее             1   
...         ...            ...        ...       ...           ...   
21489         2            NaN         47   Среднее             1   
21495         1            NaN         50   среднее             1   
21497         0            NaN         48    ВЫСШЕЕ             0   
21502         1            NaN         42   среднее             1   
21510         2            NaN         28   среднее             1   

               family_status  family_status_id gender  income_type  debt  \
12          гражданский брак                 1      M    пенсионер     0   
26           женат / замужем                 0      M  госслужащий     0   
29     Не женат / не замужем                 4      F    пенсионер     0   
41           женат / замужем                 0      F  госслужащий     0   
55          гражданский брак                 1      F    пенсионер     1   
...                      ...               ...    ...          ...   ...   
21489        женат / замужем                 0      M    компаньон     0   
21495       гражданский брак                 1      F    сотрудник     0   
21497        женат / замужем                 0      F    компаньон     0   
21502        женат / замужем                 0      F    сотрудник     0   
21510        женат / замужем                 0      F    сотрудник     0   

       total_income                           purpose  
12              NaN                   сыграть свадьбу  
26              NaN                       образование  
29              NaN  строительство жилой недвижимости  
41              NaN  сделка с подержанным автомобилем  
55              NaN                   сыграть свадьбу  
...             ...                               ...  
21489           NaN              сделка с автомобилем  
21495           NaN                           свадьба  
21497           NaN        строительство недвижимости  
21502           NaN  строительство жилой недвижимости  
21510           NaN           приобретение автомобиля  

[2174 rows x 12 columns]

Все пропуски `NaN` и они действительно встречаются исключительно в одних и тех же строках.  
Раз пропуски в одних и тех же строках, значит, это точно не полностью случайные пропуски. При этом пропуски в каждом из столбцов не зависят от других значений в этих столбцах, а значит пропуски случайны.  
Значит, можно предположить, что пропуски в обоих столбцах зависят друг от друга: ежемесячного дохода не может быть, если нет стажа. И наоборот: если нет стажа, то нет и дохода.  
Причинами этих случайных пропусков могут быть:
* ошибка выгрузки;
* человеческий фактор (оператор не занес данные о доходе или стаже).

Также можно предположить, что пропуски были из одного источника. Возможно, все клиенты с пропущенными доходами и стажем из одного банка.    
Оценим влияние пропусков на выборку.

In [7]:
#Оценим долю пропусков во всех столбцах.
stat.isna().mean()

children            0.000000
days_employed       0.100999
dob_years           0.000000
education           0.000000
education_id        0.000000
family_status       0.000000
family_status_id    0.000000
gender              0.000000
income_type         0.000000
debt                0.000000
total_income        0.100999
purpose             0.000000
dtype: float64

В каждом из двух столбцов примерно по 10% значений пропущено.  
Т.к. в столбце `days_employed` кроме пропусков есть также отрицательные значения, то заполнять пропуски до их устранения нет смысла - отрицательные значения будут портить значения среднего и медианы. Поэтому займемся этим на втором шаге, после работы с отрицальными значениями.  
Сейчас заполним пропуски в столбце `total_income`.

In [8]:
#Определим минимальную и максимальную зарплату заемщиков.
stat.agg({'total_income': ['min', 'max']})

total_income
min  2.066726e+04
max  2.265604e+06

Данные разнятся довольно сильно: от чуть выше 20 тысяч до более 2 миллионов. В дополнение к этому, пропусков более 10%.  
Поэтому корректнее всего для заполнения пропусков использовать медианную зарплату вместо средней.

In [9]:
#Заполним пропуски медианным значением
stat['total_income'] = stat['total_income'].fillna(stat['total_income'].median())
#Проверим, что пропуски заменились
stat['total_income'].isna().sum()

0

Теперь пропусков в столбце `total_income` нет: все они заменены на медианные значения.

### Проверка данных на аномалии и исправления

По обзору данных было замечено, что в столбце `days_employed` имелись отрицательные значения наряду с пропущенными.  
Устраним пропуски и аномалии в этом столбце, а также проверим, нет ли аномалий в критичных для исследования столбцах: количество детей, семейное положение, цель кредита.

In [10]:
#Посмотрим внимательнее на столбец days_employed. 
#Посчитаем количество положительных значений и найдем крайние значения среди них
stat[stat['days_employed']>0].agg({'days_employed': ['min', 'max', 'count']})

days_employed
min    328728.720605
max    401755.400475
count    3445.000000

In [11]:
#Посмотрим, сколько строк с отрицательными значениями в столбце days_employed
stat[stat['days_employed']<0]['days_employed'].count()

15906

In [12]:
#Рассчитаем, какой процент строк содержит положительные значения в столбце days_employed
stat[stat['days_employed']>0]['days_employed'].count() / len(stat) * 100

16.004645760743323

Среди неотрицательных значений стажа работы минимальное значение около 328 тысяч. Согласно документации на данные столбец стажа указан в днях. Однако, 328 тысяч для стажа в днях - слишком много. Возможно, что эти значения выгрузились не в днях, а в часах.  
Однако, также присутсвуют отрицательные значения и пропуски. Причем, их значительно больше, чем положительных: положительных значений лишь около 16% из всей выборки.  
Даже если бы положительные значения содержали значения, похожие на количество дней, то заполнять очевидно неверные значения (ниже нуля) и пропуски значениями, основанными на меньшей части выборки, неверно.  
Также, учитывая тот факт, что в дальнейшем анализе значения из этого столбца не принимают участия, удалим этот столбец целиком.


In [13]:
#Удаляем столбец days_employed
stat = stat.drop(columns='days_employed')
#Проверяем, что удалили столбец
stat.columns

Index(['children', 'dob_years', 'education', 'education_id', 'family_status',
       'family_status_id', 'gender', 'income_type', 'debt', 'total_income',
       'purpose'],
      dtype='object')

Столбец `days_employed` удален. Теперь проверим остальные колонки на аномалии. Начнем по порядку: с количества детей.

In [14]:
stat['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

Видно, что есть аномальные значения: `-1` и `20`. Иметь 20 детей в принципе реально (приемных, например), однако количество заемщиков с таким количеством детей слишком велико: их даже больше, чем заемщиков с четырьмя и пятью детьми. Возможно, при внесении данных о количестве детей была допущена опечатка. Поэтому заменим значения `20` на `2`, а значения `-1` на `1`.

In [15]:
#Меняем значения 20 и -1 в столбце children на 2 и 1 соответственно
stat.loc[stat['children']==20, 'children'] = 2
stat.loc[stat['children']==-1, 'children'] = 1
#Проверяем, что значения заменились
stat['children'].value_counts()

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

Перейдем к выявлению аномалий в столбце `family_status`. Также необходимо проверить, что нет аномалий в связанном с ним столбце `family_status_id`

In [16]:
stat['family_status'].value_counts()

женат / замужем          12380
гражданский брак          4177
Не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64

In [17]:
stat['family_status_id'].value_counts()

0    12380
1     4177
4     2813
3     1195
2      960
Name: family_status_id, dtype: int64

Пять типов значений в столбце `family_status` и пять значений в `family_status_id`. Никаких аномалий не обнаружено. Очередь за столбцом `debt` - проверим, там должны быть только нули и единицы.

In [18]:
stat['debt'].value_counts()

0    19784
1     1741
Name: debt, dtype: int64

Действительно, в столбце `debt` только нули и единицы. Остался только `purpose`.

In [19]:
stat['purpose'].value_counts()

свадьба                                   797
на проведение свадьбы                     777
сыграть свадьбу                           774
операции с недвижимостью                  676
покупка коммерческой недвижимости         664
покупка жилья для сдачи                   653
операции с жильем                         653
операции с коммерческой недвижимостью     651
покупка жилья                             647
жилье                                     647
покупка жилья для семьи                   641
строительство собственной недвижимости    635
недвижимость                              634
операции со своей недвижимостью           630
строительство жилой недвижимости          626
покупка недвижимости                      624
строительство недвижимости                620
покупка своего жилья                      620
ремонт жилью                              612
покупка жилой недвижимости                607
на покупку своего автомобиля              505
заняться высшим образованием      

В столбце `purpose` много разных значений, но все они не аномальны. Для большего удобства в представлении данных лучше разделить значения этого столбца на разные категории. Из уникальных значений можно выделить следующие категории:
* операции с автомобилем;
* операции с недвижимостью;
* проведение свадьбы;
* получение образования.

Займемся категорированием этого столбца позднее.

### Изменение типов данных

Во всех колонках типы данных соответствуют содержащимся там значениям, кроме колонки `total_income`. Удобнее оперировать доходом, опираясь на целочисленные значения. Поэтому приведем значения в этом столбце к типу `int`. 

In [20]:
#Т.к. мы знаем, что все значения в столбце float, а пропусков нет, воспользуемся astype()
stat['total_income'] = stat['total_income'].astype('int')
#Проверим тип в столбце
stat.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21525 non-null  int64 
 1   dob_years         21525 non-null  int64 
 2   education         21525 non-null  object
 3   education_id      21525 non-null  int64 
 4   family_status     21525 non-null  object
 5   family_status_id  21525 non-null  int64 
 6   gender            21525 non-null  object
 7   income_type       21525 non-null  object
 8   debt              21525 non-null  int64 
 9   total_income      21525 non-null  int32 
 10  purpose           21525 non-null  object
dtypes: int32(1), int64(5), object(5)
memory usage: 1.7+ MB


Тип данных изменен: у всех значений в столбце `total_income` тип `int`.

### Удаление дубликатов

Из просмотра верхних строк датафрейма помним, что в колонке `education` существовали неявные дубликаты. Устраним их, приведя все значения в этом столбце к нижнему регистру.

In [21]:
#Присвоим всем значениям в столбце education аналогичные значения, только в нижнем регистре
stat['education'] = stat['education'].str.lower()
#Проверим уникальные значения в столбце
stat['education'].unique()

array(['высшее', 'среднее', 'неоконченное высшее', 'начальное',
       'ученая степень'], dtype=object)

Неявные дубликаты в столбце `education` устранены. Остальные значащие в исследовании колонки мы проверяли ранее, неявных дубликатов обнаружено не было.  
Посмотрим, есть ли явные дубликаты, и удалим их при наличии.

In [22]:
#Проверим есть ли явные дубликаты и сколько их
stat.duplicated().sum()

71

In [23]:
#Удалим явные строки-дубликаты, обновив индексы в фрейме и удалив старые
stat = stat.drop_duplicates().reset_index(drop=True)
#Проверим, удалились ли дубликаты, посчитав их
stat.duplicated().sum()

0

Дубликаты удалены. Для их поиска и удаления использовались методы библиотеки `pandas` - `duplicated()` и `drop_duplicates()` в связке со стандартным методом `sum()` и методом из `pandas` - `reset_index()`.  

Наличие дубликатов, возможно, связано с ошибкой выгрузки. Также, возможно, все дублирующиеся заемщики обслуживались в одном банке, и в базе данных этого банка произошла ошибка, в результате которых данные о заемщиках продублировались.

### Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма

Для удобства обработки фрейма и будущего построения сводных таблиц, проведем декомпозицию исходного датафрейма:
* сформируем дополнительные датафреймы словарей для столбцов `education` и `family_status`;
* удалим столбцы из исходного фрейма, для которых есть словарь.

In [24]:
#Создадим новую таблицу из исходного фрейма, в которой каждому уникальному значению из education
#соответствует уникальное значение из education_id
education_dict = stat[['education', 'education_id']].drop_duplicates().reset_index(drop=True)
#Проверим созданную таблицу
education_dict

education  education_id
0               высшее             0
1              среднее             1
2  неоконченное высшее             2
3            начальное             3
4       ученая степень             4

In [25]:
#Создадим новую таблицу из исходного фрейма, в которой каждому уникальному значению из family_status
#соответствует уникальное значение из family_status_id
family_status_dict = stat[['family_status', 'family_status_id']].drop_duplicates().reset_index(drop=True)
#Проверим созданную таблицу
family_status_dict

family_status  family_status_id
0        женат / замужем                 0
1       гражданский брак                 1
2         вдовец / вдова                 2
3              в разводе                 3
4  Не женат / не замужем                 4

In [26]:
#Удалим столбцы, для которых сформировали словари (education и family_status) из исходного фрейма
stat = stat.drop(columns=['education', 'family_status'])
#Проверим, что столбцы удалились
stat.columns

Index(['children', 'dob_years', 'education_id', 'family_status_id', 'gender',
       'income_type', 'debt', 'total_income', 'purpose'],
      dtype='object')

Мы удалили лишние столбцы из исходного датафрейма. Теперь в фрейме нет излишней информации, дубликатов, для столбцов с идентификаторами есть таблицы-словари. Можно приступать к категорированию.

### Категоризация дохода

Разделим замещиков на категории согласно их доходу:
* 0–30000 — 'E';
* 30001–50000 — 'D';
* 50001–200000 — 'C';
* 200001–1000000 — 'B';
* 1000001 и выше — 'A'.

Для записи категории заемщика по доходу создадим новый столбец во фрейме `total_income_category`.

In [27]:
#Создадим функцию для категоризации дохода по указанным выше параметрам
def group_income(income):
    if 0<income<=30000: return 'E'
    elif 30000<income<=50000: return 'D'
    elif 50000<income<=200000: return 'C'
    elif 200000<income<=1000000: return 'B'
    elif income>1000000: return 'A'
#Если ни в один диапазон доход не попадает, функция вернет None по умолчанию. 
#Проверим, что функция работает правильно.
#Создадим тестовую колонку.
test_series = pd.Series(data=[-1, 0, 20, 3000, 30000, 32000, 100000, 344552, 1450000])
#Применим к ней функцию.
test_series.apply(group_income)

0    None
1    None
2       E
3       E
4       E
5       D
6       C
7       B
8       A
dtype: object

In [28]:
#Применим созданную функцию к столбцу total_income.
#Категории, которые возвращает функция запишем в новый столбец total_income_category
stat['total_income_category'] = stat['total_income'].apply(group_income)
#Проверим, что новый столбец создан и категории соответствуют значениям
stat[:5]

children  dob_years  education_id  family_status_id gender income_type  \
0         1         42             0                 0      F   сотрудник   
1         1         36             1                 0      F   сотрудник   
2         0         33             1                 0      M   сотрудник   
3         3         32             1                 0      M   сотрудник   
4         0         53             1                 1      F   пенсионер   

   debt  total_income                     purpose total_income_category  
0     0        253875               покупка жилья                     B  
1     0        112080     приобретение автомобиля                     C  
2     0        145885               покупка жилья                     C  
3     0        267628  дополнительное образование                     B  
4     0        158616             сыграть свадьбу                     C

In [29]:
#Проверим, что в колонке total_income_category нет пропусков, посмотрев уникальные значения
stat['total_income_category'].unique()

array(['B', 'C', 'D', 'E', 'A'], dtype=object)

Доходы заемщиков категоризированы. Теперь можно наглядно показывать зависимость возрата кредита от категории дохода.  
Создадим словарь для категорий дохода, а столбец с доходами из исходного фрейма удалим.

In [30]:
#Создаем таблицу-словарь для категорий доходов
total_income_dict = pd.DataFrame(
    data=[['0-30000', 'E'],
          ['30001–50000', 'D'],
          ['50001–200000', 'C'],
          ['200001–1000000', 'B'],
          ['1000001 и выше', 'A']],
    columns=['total_income', 'total_income_category'])
#Проверим создание таблицы-словаря
total_income_dict

total_income total_income_category
0         0-30000                     E
1     30001–50000                     D
2    50001–200000                     C
3  200001–1000000                     B
4  1000001 и выше                     A

In [31]:
#Удалим столбец 'total_income' из исходного фрейма
stat = stat.drop(columns='total_income')
#Проверим, что столбцы удалились
stat.columns

Index(['children', 'dob_years', 'education_id', 'family_status_id', 'gender',
       'income_type', 'debt', 'purpose', 'total_income_category'],
      dtype='object')

### Категоризация целей кредита

Разделим замещиков на категории согласно их целям кредита:
* операции с автомобилем;
* операции с недвижимостью;
* проведение свадьбы;
* получение образования.

Для выяснения по каким подстрокам относить цели к той или иной категории методом `unique()` определим уникальные значения целей. После этого напишем функцию, которая в зависимости от значения в столбце `purpose` будет возвращать категорию. Для категорий создадим новый столбец `purpose_category`.

In [32]:
#Проверим уникальные значения в столбце purpose
stat['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

In [33]:
#Создадим функцию, которая возвращает категорию в зависимости от значения в столбце purpose
def group_purpose(purpose):
    if 'авто' in purpose: return 'операции с автомобилем'
    if 'недв' in purpose or 'жил' in purpose: return 'операции с недвижимостью'
    if 'свад' in purpose: return 'проведение свадьбы'
    if 'образ' in purpose: return 'получение образования'
#Если мы ошиблись с подстроками, функция будет возвращать unknown
    return 'unknown'
#Проверим, что функция работает правильно.
#Создадим тестовую колонку.
test_series = pd.Series(data=['покупка', 'образование', 'свадьба в малиновке', 'ааааавтомобиль'])
#Применим к ней функцию.
test_series.apply(group_purpose)

0                   unknown
1     получение образования
2        проведение свадьбы
3    операции с автомобилем
dtype: object

In [34]:
#Создадим новый столбец с категориями, присвоим ему значения, которое возвращает функция group_purpose
stat['purpose_category'] = stat['purpose'].apply(group_purpose)
#Проверим создание столбца, его значения
stat[:5]

children  dob_years  education_id  family_status_id gender income_type  \
0         1         42             0                 0      F   сотрудник   
1         1         36             1                 0      F   сотрудник   
2         0         33             1                 0      M   сотрудник   
3         3         32             1                 0      M   сотрудник   
4         0         53             1                 1      F   пенсионер   

   debt                     purpose total_income_category  \
0     0               покупка жилья                     B   
1     0     приобретение автомобиля                     C   
2     0               покупка жилья                     C   
3     0  дополнительное образование                     B   
4     0             сыграть свадьбу                     C   

           purpose_category  
0  операции с недвижимостью  
1    операции с автомобилем  
2  операции с недвижимостью  
3     получение образования  
4        проведение свадьбы

In [35]:
#Проверим столбец purpose_category на пропуски (есть ли в уникальных unknown)
stat['purpose_category'].unique()

array(['операции с недвижимостью', 'операции с автомобилем',
       'получение образования', 'проведение свадьбы'], dtype=object)

Цели заемщиков категоризированы. Теперь можно наглядно показывать зависимость возрата кредита от категории цели.  
Создадим словарь для категорий целей, а столбец с целями из исходного фрейма удалим.

In [36]:
#Создаем таблицу-словарь для категорий доходов
purpose_dict = stat[['purpose', 'purpose_category']].drop_duplicates().reset_index(drop=True)
#Проверим создание таблицы-словаря
purpose_dict

purpose          purpose_category
0                            покупка жилья  операции с недвижимостью
1                  приобретение автомобиля    операции с автомобилем
2               дополнительное образование     получение образования
3                          сыграть свадьбу        проведение свадьбы
4                        операции с жильем  операции с недвижимостью
5                              образование     получение образования
6                    на проведение свадьбы        проведение свадьбы
7                  покупка жилья для семьи  операции с недвижимостью
8                     покупка недвижимости  операции с недвижимостью
9        покупка коммерческой недвижимости  операции с недвижимостью
10              покупка жилой недвижимости  операции с недвижимостью
11  строительство собственной недвижимости  операции с недвижимостью
12                            недвижимость  операции с недвижимостью
13              строительство недвижимости  операции с недвижимостью
14      на покупку подержанного автомобиля    операции с автомобилем
15            на покупку своего автомобиля    операции с автомобилем
16   операции с коммерческой недвижимостью  операции с недвижимостью
17        строительство жилой недвижимости  операции с недвижимостью
18                                   жилье  операции с недвижимостью
19         операции со своей недвижимостью  операции с недвижимостью
20                              автомобили    операции с автомобилем
21                   заняться образованием     получение образования
22        сделка с подержанным автомобилем    операции с автомобилем
23                   получение образования     получение образования
24                              автомобиль    операции с автомобилем
25                                 свадьба        проведение свадьбы
26   получение дополнительного образования     получение образования
27                    покупка своего жилья  операции с недвижимостью
28                операции с недвижимостью  операции с недвижимостью
29           получение высшего образования     получение образования
30                         свой автомобиль    операции с автомобилем
31                    сделка с автомобилем    операции с автомобилем
32                  профильное образование     получение образования
33                      высшее образование     получение образования
34                 покупка жилья для сдачи  операции с недвижимостью
35                   на покупку автомобиля    операции с автомобилем
36                            ремонт жилью  операции с недвижимостью
37            заняться высшим образованием     получение образования

In [37]:
#Удалим столбец 'purpose' из исходного фрейма
stat = stat.drop(columns='purpose')
#Проверим, что столбцы удалились
stat.columns

Index(['children', 'dob_years', 'education_id', 'family_status_id', 'gender',
       'income_type', 'debt', 'total_income_category', 'purpose_category'],
      dtype='object')

## Построение зависимостей

### Количество детей и возврат кредита в срок

Для проверки наличия зависимости между возвратом кредита и количеством детей построим сводную таблицу, применив функцию `pivot` к столбцу `children`. В таблице будет отображаться количество заемщиков, сгрупированных по количеству детей, отдавшие или нет кредит вовремя.

In [38]:
# Общая функция построения сводной таблицы
def pivot(column):
    return (stat.pivot_table(index=column, 
                            values='debt', 
                            aggfunc=['count', 'sum', 'mean'])
                .sort_values(by=('mean', 'debt'))
                .rename(columns={'mean': 'доля просрочек',
                                 'count': 'число заемщиков',
                                 'sum': 'число просрочек'})
                .droplevel(1, axis=1))

In [39]:
children = pivot('children')
children

число заемщиков  число просрочек  доля просрочек
children                                                  
5                       9                0        0.000000
0                   14091             1063        0.075438
3                     330               27        0.081818
1                    4855              445        0.091658
2                    2128              202        0.094925
4                      41                4        0.097561

В группе заемщиков с пятью детьми всего 9 человек. И из них ни один не имел задолженностей.  
В остальных группах заемщиков по детям процент задолженности минимален у заемщиков без детей - 7,54%.  
Максимальный процент невозвратов в срок у заемщиков с четырьмя детьми - 9,76%.


**Вывод 1**

По полученным данным можно сделать вывод, что в категории заемщиков с детьми доля просрочек больше, чем у заемщиков без детей.

### Семейное положение и возврат кредита в срок

Для проверки наличия зависимости между возвратом кредита и семейным положением построим сводную таблицу, применив функцию `pivot` к столбцу `family_status_id`. В таблице будет отображаться количество заемщиков, сгрупированных по семейному положению, отдавшие или нет кредит вовремя.  
Т.к. в столбце `family_status_id` хранятся только идентификаторы семейного положения, добавим к сводной таблице соответствующие идентификаторам значения из словаря `family_status_dict`.

In [40]:
#Отображение сводной таблицы для групп заемщиков по семейному положению.
#Методом merge объединяем сводную таблицу со словарем по столбцу family_status_id.
pivot('family_status_id').merge(family_status_dict, on='family_status_id')

family_status_id  число заемщиков  число просрочек  доля просрочек  \
0                 2              959               63        0.065693   
1                 3             1195               85        0.071130   
2                 0            12339              931        0.075452   
3                 1             4151              388        0.093471   
4                 4             2810              274        0.097509   

           family_status  
0         вдовец / вдова  
1              в разводе  
2        женат / замужем  
3       гражданский брак  
4  Не женат / не замужем

Наибольший процент просрочек кредита у заемщиков, которые никогда не состояли в браке - 9,75%.  
Наименьшие проблемы с возвратом кредита в срок имеют люди, бывшие в браке, но в данный момент в нем не находящиеся:
* у вдов/вдовцов процент просрочек - 6,57%;
* у разведенных людей - 7,11%.

Люди, состоящие в официальных отношениях отдают долг с просрочкой примерно в таком же проценте случаев, что и разведенные - 7,55%.  
Люди в гражданском браке недалеко ушли от никогда не состоявших в браке - 9,35%.

**Вывод 2**

Наименее способными отдавать в срок кредит людьми оказались неженатые/незамужние (9,75% просрочек) и люди в гражданском браке (9,35% просрочек).  
У людей, находящихся в официальном браке или бывшем в нем, процент просрочек заметно ниже:
* у женатых/замужних - 7,55%;
* у разведенных - 7,11%;
* у вдов/вдовцов - 6,57%.

### Уровень дохода и возврат кредита в срок

Для проверки наличия зависимости между возвратом кредита и уровнем дохода построим сводную таблицу, применив функцию `pivot` к столбцу `total_income_category`. В таблице будет отображаться количество заемщиков, сгрупированных по уровню дохода, отдавшие или нет кредит вовремя.  
Т.к. в столбце `total_income_category` хранятся только буквенное соответствие уровня дохода, добавим к сводной таблице соответствующие буквам значения из словаря `total_income_dict`.

In [41]:
#Отображение сводной таблицы для групп заемщиков по уровню дохода.
#Методом merge объединяем сводную таблицу со словарем по столбцу total_income_category.
pivot('total_income_category').merge(total_income_dict, on='total_income_category')

total_income_category  число заемщиков  число просрочек  доля просрочек  \
0                     D              350               21        0.060000   
1                     B             5041              356        0.070621   
2                     A               25                2        0.080000   
3                     C            16016             1360        0.084915   
4                     E               22                2        0.090909   

     total_income  
0     30001–50000  
1  200001–1000000  
2  1000001 и выше  
3    50001–200000  
4         0-30000

Наибольший процент просрочек кредита у заемщиков с минимальным доходом (до 30000) - 9,09%. Однако выборка очень мала: таких заемщиков лишь 22.  
8,49% заемщиков со средним доходом (от 50001 до 200000) не возвращают кредит во время.  
Среди клиентов с максимальным доходом (выше 1000001) 8% не укладываются в срок возврата. Однако, как и в случае с клиентами с минимальными доходом выборка маленькая: только 25 человек.  
Среди клиентов с доходом выше среднего (от 200001 до 1000000) 7,06% с просрочками по кредиту, а среди клиентов с доходом ниже среднего (от 30001 до 50000) таких меньше 6%.

**Вывод 3**

Если рассматривать репрезентативные выборки с числом заемщиков в категории от 100 человек и выше, то зависимости между уровнем дохода и возвратом кредита вовремя не наблюдается. Кредит вовремя не возвращают:
* 8,49% заемщиков со средним доходом;
* 7,06% заемщиков с доходом выше среднего;
* 6% заемщиков с доходом ниже среднего.

Вопреки ожиданиям, самыми надежными в плане соблюдения сроков возврата кредита оказались заемщики с доходом ниже среднего.

### Цель кредита и возврат кредита в срок

Для проверки наличия зависимости между возвратом кредита и целью кредита построим сводную таблицу, применив функцию `pivot` к столбцу `purpose_category`. В таблице будет отображаться количество заемщиков, сгрупированных по цели кредита, отдавшие или нет кредит вовремя.

In [42]:
#Отображение сводной таблицы для групп заемщиков по цели кредита
pivot('purpose_category')

число заемщиков  число просрочек  доля просрочек
purpose_category                                                          
операции с недвижимостью            10811              782        0.072334
проведение свадьбы                   2324              186        0.080034
получение образования                4013              370        0.092200
операции с автомобилем               4306              403        0.093590

В каждой категории достаточно большое количество заемщиков, чтобы говорить о полноте выборки.  
Наибольшие проблемы с возвратом в срок возникают у заемщиков при цели кредита, связанной с автомобилем - 9,36% заемщиков возвращают не в срок.  
Невовремя возвращают 9,22% кредитов, полученных на образование.  
Кредиты, взятые на свадьбу, не возвращают вовремя в 8% случаев.  
Наименее подвержены невозврату вовремя кредиты, взяты на недвижимость - 7,23% кредитов с просрочками.

**Вывод 4**

Среди пары цель кредита и его возврат в срок сложно говорить о зависимости, т.к. группы (цели кредита) по сути не связаны между собой.  
Однако, можно выделить наиболее и наименее подверженные риску невозврата вовремя цели кредитов:
* меньше всего просрочек у кредитов с целями, связанными с недвижимостью - 7,23%;
* 8% кредитов на свадьбы вернутся не в срок;
* почти одинаковый процент кредитов, возвращенных не в срок, у кредитов на образование (9,22%) и на операции с авто (9,36%).

## Общий вывод

Проанализирована статистика платежеспособности клиентов банка.  
Была произведена предобработка данных, заполнены пропуски, исправлены аномалии, удалены дубликаты.  
Была произведена категоризация доходов и целей кредита, созданы таблицы-словари для семейного положения, доходов, целей кредита.  
Построены сводные таблицы зависимостей возврата кредита в срок и:
* количества детей у заемщика;
* семейного положения заемщика;
* уровня дохода;
* цели кредита.

1. Гипотеза о том, что с увеличением количества детей растет процент просрочек кредита подтверждена (доля просрочек у заемщиков с детьми выше, чем у заемщиков без детей).
2. Наименее способными отдавать в срок кредит людьми оказались люди, которые никогда не были в официальном браке (9,75% и 9,35% просрочек). У бывших или находящихся в браке ответственность к сроку выплаты кредита больше: на 2% меньше просрочек.
3. В некоторых категориях уровня дохода заемщиков выборка оказалась слишком мала, чтобы говорить о закономерностях. По остальным группам дохода зависимости нет: наименьший процент просрочек у людей с доходом ниже среднего. В добавок к этому в данном столбце было много пропущенных значений (10% от общего числа), что, возможно, повлияло на результат.
4. Среди целей кредита самыми подверженными просрочкам оказались операции с авто и образование (9,36% и 9,22% соответственно). Кредиты, взятые на недвижимость наименее подвержены просрочкам - лишь 7,23% невовремя возвращенных кредитов. Кредиты на свадьбы имеют 8% просрочек среди всех взятых на свадьбы кредитов.

Рекомендуется:
* проверить базу на способ хранения стажа заемщика - подавляющее большинство значений в этом столбце хранится в несоответствующем описанию формате;
* в пояснительную записку к данным добавить расшифровку колонки `debt` (0 - кредит возвращен вовремя, 1 - с просрочкой);
* для столбца `education` заменить ручной ввод значения на выбор значения из выпадающего списка для унификации записываемых значений;
* провести исследование о зависимости уровня образования, возраста, пола и способности возврата кредита вовремя. Все необходимые данные уже есть.